gprofiler/metadata/system_metadata.py

```diff
@@ -0,0 +1,218 @@
+import array
+import fcntl
+import os
+import platform
+import re
+import socket
+import struct
+import subprocess
+import sys
+import time
+from dataclasses import dataclass
+from typing import Optional, Tuple
+
+import distro  # type: ignore
+import psutil
+
+from gprofiler.log import get_logger_adapter
+
+logger = get_logger_adapter(__name__)
+hostname: Optional[str] = None
+
+
+def get_libc_version() -> Tuple[str, str]:
+    # platform.libc_ver fails for musl, sadly (produces empty results).
+    # so we'll run "ldd --version" and extract the version string from it.
+    # not passing "encoding"/"text" - this runs in a different mount namespace, and Python fails to
+    # load the files it needs for those encodings (getting LookupError: unknown encoding: ascii)
+    def decode_libc_version(version: bytes) -> str:
+        return version.decode("utf-8", errors="replace")
+
+    # Avoid circular imports
+    from gprofiler.utils import run_process
+
+    ldd_version = run_process(
+        ["ldd", "--version"], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, suppress_log=True, check=False
+    ).stdout
+    # catches GLIBC & EGLIBC
+    m = re.search(br"GLIBC (.*?)\)", ldd_version)
+    if m is not None:
+        return "glibc", decode_libc_version(m.group(1))
+    # catches GNU libc
+    m = re.search(br"\(GNU libc\) (.*?)\n", ldd_version)
+    if m is not None:
+        return "glibc", decode_libc_version(m.group(1))
+    # musl
+    m = re.search(br"musl libc.*?\nVersion (.*?)\n", ldd_version, re.M)
+    if m is not None:
+        return "musl", decode_libc_version(m.group(1))
+
+    return "unknown", decode_libc_version(ldd_version)
+
+
+def get_run_mode_and_deployment_type() -> Tuple[str, str]:
```

Comment:
I think it's nicer to keep these 2 logics separate - `get_run_mode` as one, and another:
```
def get_deployment_type(run_mode: str):
   if run_mode == k8s ..
   elif ...
....
```